### PR TITLE
Allow colon in mlflow param and metric names

### PR DESCRIPTION
**Pull request by piyushdaftary**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/piyushdaftary/mlflow/pull/12943?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12943/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12943
```

</p>
</details>

### Related Issues/PRs

https://github.com/mlflow/mlflow/issues/12940

### What changes are proposed in this pull request?

Adding Support for colon (`:`) in mlflow param and metric names. 
Updated validation unit test to support colon.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [x] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [x] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by piyushdaftary** [comment]
@BenWilson2 @dbczumar  : Please review the PR. Thanks

File: tests/utils/test_validation.py [hunk 1]
```diff
@@ -26,6 +26,7 @@
     "a/b/c",
     "a.b.c",
     ".a",
+    ":",
```
Comment (WeichenXu123) [inline tests/utils/test_validation.py:29]: could you add test name like `aa:bb:cc` too ?

**Comment by WeichenXu123** [comment]
According to use-case https://github.com/mlflow/mlflow/issues/12940#issuecomment-2289414816 , I think this change is good, CC @mlflow-automation to see if there are other concerns.

**Review by daniellok-db — APPROVED** [review]
lgtm pending https://github.com/mlflow/mlflow/pull/12943/files#r1722637407

File: tests/utils/test_validation.py (hunk 1 above)
Comment (harupy) [reply tests/utils/test_validation.py:29]:
```suggestion
    ":",
    "aa:bb:cc",
```

File: mlflow/utils/validation.py [hunk 2]
```diff
@@ -36,7 +36,7 @@
 
 _BAD_CHARACTERS_MESSAGE = (
     "Names may only contain alphanumerics, underscores (_), dashes (-), periods (.),"
-    " spaces ( ), and slashes (/)."
+    " spaces ( ), colon(:) and slashes (/)."
```
Comment (harupy) [inline mlflow/utils/validation.py:39]: On windows, can files and directory names contain `:`?

File: mlflow/utils/validation.py [hunk 3]
```diff
@@ -18,7 +18,7 @@
 
 # Regex for valid param and metric names: may only contain slashes, alphanumerics,
 # underscores, periods, dashes, and spaces.
```
Comment (daniellok-db) [inline mlflow/utils/validation.py:20]:
```suggestion
# underscores, periods, dashes, colons, and spaces.
```

**Comment by WeichenXu123** [comment]
Quick polling: @piyushdaftary is asking if this change can go into next patching release. I think it is safe to be added to next patching release.

File: mlflow/utils/validation.py (hunk 2 above)
Comment (WeichenXu123) [reply mlflow/utils/validation.py:39]: good point. But Linux path does not allow "/" char too, why does it work? or is it not covered by test ?

Comment (WeichenXu123) [reply mlflow/utils/validation.py:39]: @piyushdaftary 
Good point . Could you add another test to log param / metric with key name including these good characters?

Comment (WeichenXu123) [reply mlflow/utils/validation.py:39]: Windows doesn't support `:`, so in windows, bad char list should still contain colon char.

Comment (piyushdaftary) [reply mlflow/utils/validation.py]: Good Point. I have refactored the code to handle `colon` for windows and non-windows platform differently. Please have a look. @WeichenXu123 @harupy

**Review by BenWilson2 — APPROVED** [review]
LGTM!

**Review by WeichenXu123 — APPROVED** [review]
LGTM

File: mlflow/utils/validation.py [hunk 4]
```diff
@@ -111,6 +103,34 @@ def bad_path_message(name):
...
+            "Names may only contain alphanumerics, underscores (_), dashes (-), periods (.),"
+            " spaces ( ) and slashes (/)."
+        )
+
+    # For other system valid param and metric names: may only contain slashes, alphanumerics,
+    # underscores, periods, dashes, colons, and spaces.
+    return (
+        "Names may only contain alphanumerics, underscores (_), dashes (-), periods (.),"
+        " spaces ( ), colon(:) and slashes (/)."
+    )
```
Comment (harupy) [inline mlflow/utils/validation.py:131]: /ai simplify this function.

Comment (harupy) [reply mlflow/utils/validation.py:131]:
```suggestion
def bad_character_message():
    ... comment
    msg = (
        "Names may only contain alphanumerics, underscores (_), dashes (-), periods (.),"
        " spaces ( ){} and slashes (/)."
    )
    return msg.format(", colon(:)") if is_windows() else msg.format("")
```

**Review by BenWilson2 — APPROVED** [review]

**Pull request merged**